website/docs/BestPractices.md

```diff
@@ -0,0 +1,203 @@
+
+Queries all table rows.
+
+```js
+// 👎
+await $$('table tr')[15];
+```
+
+Queries a single table row.
+
+```js
+// 👍
+await $('table tr:nth-child(15)');
+```
+
+## Use the built-in assertions
+
+Don't use manual assertions that do not automatically wait for the results to match as this will cause for flaky tests.
+
+```js
+// 👎
+expect(await button.isDisplayed()).toBe(true);
+```
+
+By using the built-in assertions WebdriverIO will automatically wait for the actual result to match the expected result, resulting in resilient tests.
+It achieves this by automatically retrying the assertion until it passes or times out.
+
+```js
+// 👍
+await expect(button).toBeDisplayed();
+```
+
+## Don't overuse commands and assertions
+
+When using expect.toBeDisplayed you implicitly also wait for the element to exist. There isn't a need to use the waitForXXX commands when you already have an assertion doing the same thing.
+
+```js
+// 👎
+await button.waitForExist();
+await expect(button).toBeDisplayed();
+
+// 👎
+await button.waitForDisplayed();
+await expect(button).toBeDisplayed();
+
+// 👍
+await expect(button).toBeDisplayed();
+```
+
+No need to wait for an element to exist or be displayed when interacting or when asserting something like it's text unless the element can explicitly be invisible (opacity: 0; for example) or can explicitly be disabled (disabled attribute for example) in which case waiting for the element to be displayed makes sense.
+
+```js
+// 👎
+await expect(button).toBeExisting();
+await expect(button).toHaveText('Submit');
+
+// 👎
+await expect(button).toBeDisplayed();
+await expect(button).toHaveText('Submit');
+
+// 👎
+await expect(button).toBeDisplayed();
+await expect(button).click();
+```
+
+```js
+// 👍
+await expect(button).click();
+
+// 👍
+await expect(button).toHaveText('Submit');
+```
+
+## Lint your code
+
+Using eslint to lint your code you can potentionally catch errors early, use our [linting rules](https://www.npmjs.com/package/eslint-plugin-wdio) to make sure that some of the best practices are always applied.
+
+## Don't pause
+
+It can be tempting to use the pause command but using this is a bad idea as it isn't resilient and will only cause for flaky tests in the long run.
+
+```js
+// 👎
+await nameInput.setValue('Bob');
+await browser.pause(200); // wait for submit button to enable
+await submitFormButton.click();
+
+// 👍
+await nameInput.setValue('Bob');
+await submitFormButton.waitForEnabled();
+await submitFormButton.click();
+```
+
+## Async loops
+
+When you have some asynchronous code that you want to repeat, it is important to know that not all loops can do this.
+
+The following will not click the elements in order.
+
+```js
+// 👎
+for (let i = 0; i < 10; i++) {
+    await executeSomeLogic();
+}
+```
+
+The following will click the elements in order.
+
+```js
+// 👍
+for (const element of elements) {
+    await executeSomeLogic();
+}
+```
```

Comment:
I don't understand this example. There is no click happening. Also the behavior is the same, it is just a different loop.

Reply:
As shown in my example earlier the execution is not the same.
You can replicate this using browser.keys, it will run differently with the for...of loop compared to other loops.

Reply:
Yes, this I do understand because you have a for..of loop and a forEach one but the PR suggests two for loops that have the same effect.

Reply:
forEach doesn't print chars in expected order as I believe `forEach` expects sync code, so async/await has no effect when forEach is used! i don't recommend use of foeEach with async/awaits!